### PR TITLE
Scaffold `qp.allocate` interface for Phase Gradient

### PR DESCRIPTION
**Pull request by JerryChen97**
**Context:**
This PR adds the minimal frontend scaffold for phase-gradient allocation: a distinct `state="phase-grad"` allocation state plus required `precision` metadata preserved through queuing, capture, and tape construction. It intentionally does not yet implement phase-gradient preparation or lowering, and instead draws a clear error boundary for unresolved execution paths.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-116441]

### CODE REVIEW

File: pennylane/transforms/resolve_dynamic_wires.py [hunk 1]
```diff
@@ -37,6 +37,12 @@ def __init__(self, zeroed=(), any_state=(), min_int=None, allow_resets: bool = T
 
     def _retrieval_method(self, state: AllocateState):
         _retrieval_map = {AllocateState.ZERO: self._get_zeroed, AllocateState.ANY: self._get_any}
+        if state is AllocateState.PHASE_GRAD:
+            raise AllocationError(
+                "Phase-gradient allocation resolution is not yet implemented. "
+                "Concrete preparation and wire scheduling for 'phase-grad' state "
+                "will be added in a follow-up change."
```
Comment (Qottmann) [inline pennylane/transforms/resolve_dynamic_wires.py:44]: perhaps a bit too optimistic, I think there is still lots to be figured out, and thats fine, we dont technically _need_ this all too soon, or at least I'd be in favour of thinking of a really good solution, rather than a quick one :)

File: pennylane/allocation.py [hunk 2]
```diff
@@ -225,6 +256,11 @@ def allocate(
             ``restored=False`` indicates that the user does not promise to restore the dynamically
             allocated wires before being deallocated. The default value is ``False``.
 
+        precision (float | None):
+            Target precision for ``"phase-grad"`` allocations. Required when
+            ``state="phase-grad"``; must be a positive float. Must be ``None``
+            for ``"zero"`` and ``"any"`` states. The default value is ``None``.
```
Comment (Qottmann) [inline pennylane/allocation.py:262]: we might want a good default even for "phase-grad" states.
personally I think that cases where the user wants to control the precision of the phase gradient state are rather exotic, but we are not in full agreement on that point inside our team (yet) 😅

File: pennylane/allocation.py [hunk 3]
```diff
@@ -365,15 +401,34 @@ def circuit():
         1: ──H──┤↗│  │0⟩──X─┤
     """
     state = AllocateState(state)
+
+    # Validate precision constraints
+    if precision is not None and not state.is_phase_gradient():
+        raise ValueError(
```
Comment (Qottmann) [inline pennylane/allocation.py:407]: would be enough to raise a warning here as the precision argument will just be ignored no?

File: tests/test_allocation.py [hunk 4]
```diff
@@ -366,3 +376,203 @@ def c():
...
+            with qml.allocate(2, state="phase-grad", precision=1e-6, restored=True) as wires:
+                qml.CNOT((0, wires[0]))
+                qml.CNOT((wires[0], wires[1]))
+            return qml.expval(qml.Z(0))
+
+        # At the user level, the tape should contain the Allocate op with phase-grad metadata
+        tape = qml.workflow.construct_tape(circuit)()
+        alloc_ops = [op for op in tape.operations if isinstance(op, Allocate)]
+        assert len(alloc_ops) == 1
+        assert alloc_ops[0].state is AllocateState.PHASE_GRAD
```
Comment (Qottmann) [inline tests/test_allocation.py:424]: just curious: the `AllocateState` is kinda just a placeholder right? What does it do/imply in the other cases? Are there associated protocols that prepare the appropriate state (e.g. PLUS pr ONE? [not sure if they even exist])

File: tests/test_allocation.py [hunk 5]
```diff
@@ -366,3 +376,203 @@ def c():
...
+        @qml.qnode(qml.device("default.qubit", wires=3))
+        def circuit():
+            with qml.allocate(1, state="phase-grad", precision=1e-6, restored=True) as wires:
+                qml.X(wires[0])
+            return qml.expval(qml.Z(0))
+
+        with pytest.raises(qml.exceptions.AllocationError):
+            circuit()
+
+    def test_allocate_phase_grad_returns_register(self):
```
Comment (Qottmann) [inline tests/test_allocation.py]: is the distinction in the test name between test_allocate and test_qml_allocate intentional?

File: tests/test_allocation.py [hunk 6]
```diff
@@ -366,3 +376,203 @@ def c():
...
+        assert alloc_ops[0].precision == 1e-6
+        assert alloc_ops[0].is_phase_gradient
+
+    def test_qml_allocate_phase_grad_device_execution_raises(self):
+        """Test that device execution of a phase-grad allocation raises AllocationError.
+
+        Phase-gradient wire resolution is not yet implemented, so lowering to
+        device level should produce a clear error rather than silently miscompiling.
+        The inner AllocationError from resolve_dynamic_wires is wrapped by device
+        preprocessing into a higher-level message.
```
Comment (Qottmann) [inline tests/test_allocation.py:434]: whats the message? why not test it explicitly here?

File: tests/test_allocation.py [hunk 7]
```diff
@@ -366,3 +376,203 @@ def c():
...
+            allocate(2, state="phase-grad", precision=1e-6)
+
+        op = q.queue[0]
+        assert op.state is AllocateState.PHASE_GRAD
+        assert op.precision == 1e-6
+        assert op.is_phase_gradient
+
+    def test_allocate_phase_grad_precision_required(self):
+        """Test that phase-grad without precision raises ValueError."""
+        with pytest.raises(ValueError, match="precision is required for state='phase-grad'"):
```
Comment (Qottmann) [inline tests/test_allocation.py:469]: perhaps remove this and provide a default value

File: tests/test_allocation.py [hunk 8]
```diff
@@ -366,3 +376,203 @@ def c():
...
+            allocate(2, state="phase-grad")
+
+    def test_allocate_phase_grad_hyperparameters(self):
+        """Test full hyperparameters dict for phase-grad allocation."""
+        with qml.queuing.AnnotatedQueue() as q:
+            allocate(1, state="phase-grad", precision=0.001, restored=True)
+
+        op = q.queue[0]
```
Comment (Qottmann) [inline tests/test_allocation.py:477]: is "allocate" the op here? 🤔 🤯

File: tests/test_allocation.py [hunk 9]
```diff
@@ -366,3 +376,203 @@ def c():
...
+        op = Allocate.from_num_wires(1)
+        assert op.precision is None
+
+
+class TestPhaseGradValidation:
+    """Tests for precision validation rules."""
+
+    def test_precision_required_for_phase_grad(self):
+        """Test that phase-grad without precision raises ValueError."""
+        with pytest.raises(ValueError, match="precision is required for state='phase-grad'"):
```
Comment (Qottmann) [inline tests/test_allocation.py:516]: same here

File: tests/test_allocation.py [hunk 10]
```diff
@@ -366,3 +376,203 @@ def c():
...
+    def test_capture_phase_grad_state(self):
+        """Test that phase-grad state is captured in JAXPR params."""
+        import jax
+
+        def f():
+            w = allocate(2, state="phase-grad", precision=1e-6, restored=True)
+            deallocate(w)
+
+        jaxpr = jax.make_jaxpr(f)()
+        assert jaxpr.eqns[0].primitive == allocate_prim
```
Comment (Qottmann) [inline tests/test_allocation.py:557]: whats `allocate_prim` and where does it come from?

**Review by Qottmann — CHANGES_REQUESTED** [review]
Very nice @JerryChen97 !
My main feedback is that we should provide a good default value of precision instead of forcing a user to provide it. (For the moment can be 1e-6, with a _really_ good default value tbd later).

Other than that mostly questions from my side as I am not too familiar with the allocation system. If you're short on time feel free to ignore (:

**Comment by JerryChen97** [comment]
> Very nice @JerryChen97 ! My main feedback is that we should provide a good default value of precision instead of forcing a user to provide it. (For the moment can be 1e-6, with a _really_ good default value tbd later).
> 
> Other than that mostly questions from my side as I am not too familiar with the allocation system. If you're short on time feel free to ignore (:

Actually me neither super familiar with `allocate` :sweat: so no worry about questions; actually I do think the more the better :laughing:

File: pennylane/allocation.py (hunk 2 above)
Comment (JerryChen97) [reply pennylane/allocation.py:262]: I see...

File: tests/test_allocation.py (hunk 4 above)
Comment (JerryChen97) [reply tests/test_allocation.py:424]: Not a placeholder — `AllocateState` drives wire scheduling in `resolve_dynamic_wires._WireManager`. For ZERO and ANY, the wire manager sources from the appropriate pool (zeroed or any-state), so no explicit prep is needed. PHASE_GRAD is the first state requiring actual preparation gates.
PLUS/ONE don't currently exist but the enum is extensible.

File: tests/test_allocation.py (hunk 10 above)
Comment (JerryChen97) [reply tests/test_allocation.py:557]: `allocate_prim` is a JAX primitive (`QmlPrimitive`) defined in `pennylane/allocation.py`. What we're testing here is the **capture pathway**: when `qml.capture.enable()` is active, `qml.allocate()` dispatches to `allocate_prim.bind(...)` instead of queuing an `Allocate` operator on the tape. These tests verify that the phase-gradient metadata (`state` enum + `precision` value) survives the JAX tracing round-trip — i.e., that the captured JAXPR correctly records the allocation parameters. This is the capture-side counterpart to the tape-based hyperparameters tests above.

**Comment by JerryChen97** [comment]
According to feedback, the existence of input precision is in debate, therefore renamed the title to reflect this

**Comment by Qottmann** [comment]
> According to feedback, the existence of input precision is in debate, therefore renamed the title to reflect this

Sorry, I didn't mean we shouldn't add it, we should!
It's more that we need a good standard as most users that just want to run algos probably won't be touching it, but we definitely want it for research purposes (as in researchers in general, not just PL research team :))

So the main point is that precision should have a default value and should not be mandatory for the user to provide

**Comment by JerryChen97** [comment]
> > According to feedback, the existence of input precision is in debate, therefore renamed the title to reflect this
> 
> Sorry, I didn't mean we shouldn't add it, we should! It's more that we need a good standard as most users that just want to run algos probably won't be touching it, but we definitely want it for research purposes (as in researchers in general, not just PL research team :))
> 
> So the main point is that precision should have a default value and should not be mandatory for the user to provide

Ah I see. Sorry for bringing confusion! Let me adjust the current scaffold soon

**Comment by JerryChen97** [comment]
Temporarily de-prioritized for now; will come back sooon when more capacity
